extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/AzureTableStore.java

```diff
@@ -0,0 +1,285 @@
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public void updateJob(UUID jobId, PortabilityJob job) throws IOException {
+    updateJob(jobId, job, null);
+  }
+
+  @Override
+  public void updateJob(UUID jobId, PortabilityJob job, JobUpdateValidator validator)
+      throws IOException {
+    Preconditions.checkNotNull(jobId, "Job not persisted");
+    Preconditions.checkNotNull(job, "Job is null");
+    try {
+
+      CloudTable table = tableClient.getTableReference(JOB_TABLE);
+
+      String serializedJob = configuration.getMapper().writeValueAsString(job);
+      DataWrapper wrapper =
+          new DataWrapper(
+              configuration.getPartitionKey(),
+              jobId.toString(),
+              job.jobAuthorization().state().name(),
+              serializedJob);
+
+      if (validator != null) {
+        PortabilityJob previousJob = findJob(jobId);
+        if (previousJob == null) {
+          throw new IOException("Could not find record for jobId: " + jobId);
+        }
+
+        validator.validate(previousJob, job);
+      }
+
+      TableOperation insert = TableOperation.insertOrReplace(wrapper);
+      table.execute(insert);
+
+    } catch (JsonProcessingException | StorageException | URISyntaxException e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public PortabilityJob findJob(UUID jobId) {
+    Preconditions.checkNotNull(jobId, "Job not persisted");
+    try {
+
+      CloudTable table = tableClient.getTableReference(JOB_TABLE);
+
+      TableOperation retrieve =
+          TableOperation.retrieve(
+              configuration.getPartitionKey(), jobId.toString(), DataWrapper.class);
+      TableResult result = table.execute(retrieve);
+      DataWrapper wrapper = result.getResultAsType();
+      return configuration.getMapper().readValue(wrapper.getSerialized(), PortabilityJob.class);
+    } catch (StorageException | URISyntaxException | IOException e) {
+      throw new MicrosoftStorageException(e);
+    }
+  }
+
+  @Override
+  public void remove(UUID jobId) throws IOException {
+    Preconditions.checkNotNull(jobId, "Job not persisted");
+    remove(jobId, JOB_TABLE);
+  }
+
+  @Override
+  public <T extends DataModel> void create(UUID jobId, T model) {
+    try {
+      create(jobId, JOB_DATA_TABLE, null, model);
+    } catch (IOException e) {
+      throw new MicrosoftStorageException(e);
+    }
+  }
+
+  @Override
+  public <T extends DataModel> void update(UUID jobId, T model) {}
+
+  @Override
+  public <T extends DataModel> T findData(Class<T> type, UUID jobId) {
+    return find(type, jobId, JOB_DATA_TABLE);
+  }
+
+  @Override
+  public void removeData(UUID jobId) {
+    try {
+      remove(jobId, JOB_DATA_TABLE);
+    } catch (IOException e) {
+      throw new MicrosoftStorageException("Unable to remove data for job: " + jobId);
+    }
+  }
+
+  @Override
+  public void create(UUID jobId, String key, InputStream stream) {
+    try {
```

Comment:
can there be any race conditions here? We use a db transaction for the Google store, maybe something similar can be done here?

Same thing in updateJob

Reply:
I doubt there is a race condition here since the combination of jobId and key are unique and the behavior is idempotent if the creation happens twice .

Reply:
Oh ok in that case, if Cosmos doesn't support transactions for this case I agree we need to redesign. 

Another option is use a different type of hosted storage that acts more like a queue, where the store itself can be responsible for ensuring each entity is only given out to one consumer. Maybe Pubsub?

Another option, I'm not sure if this is possible but if we can enforce "final" state on the encryption key fields in the db, maybe the db operation could fail when you try and update a job's keys when it already has non-null keys assigned (from another worker).

Reply:
One way to implement the last scenario you mentioned is with a "compare-and-set" style operation. I looked into this approach (Cassandra has something similar) and discovered Cosmos has a way of doing this.

The compare is done against the expected value of something. Both workers read, then compare-and-set. One will fail and go away while the other proceeds.

Failure recovery will be an interesting case. We'll have to decide whether duplication is tolerable. For example, when a worker dies, does the transfer start over or pick up from the last known item? If it's the latter, some form of checkpointing would be required.

I'll look into the compare-and-set once I have everything containerized on Azure.

Reply:
The failure recovery will be an interesting issue. I wonder if storing the key on the client will present an issue since users  could click away before the failure occurs or an interruption could occur.

If we externalize the key creation like we discussed in removing JobMetadata that will make the other scenario you mentioned easier.